src/pybind/mgr/progress/module.py

```diff
@@ -301,23 +335,25 @@ def notify(self, notify_type, notify_data):
             for ev_id, ev in self._events.items():
                 if isinstance(ev, PgRecoveryEvent):
                     ev.pg_update(data, self.log)
-        else:
-            self.log.debug("Ignore notification '{0}'".format(notify_type))
-            pass
+                    self.maybe_complete(ev)
 
-    def _persist(self):
+    def maybe_complete(self, event):
+        if event.progress >= 1.0:
+            self._complete(event)
+
+    def _save(self):
         self.log.info("Writing back {0} completed events".format(
             len(self._completed_events)
         ))
         # TODO: bound the number we store.
         encoded = json.dumps({
-            "events": [ev.encode() for ev in self._completed_events],
+            "events": [ev.to_json() for ev in self._completed_events],
```

Comment:
The JSON encoder for custom objects (e.g. Event in this case) works out pretty well to simplify things. Here is an example https://github.com/ceph/ceph/pull/23497/files#diff-6cb378091f67c72c18f8b31a5fe5a82fR13